File: app/src/main/java/com/chess/cryptobot/task/CoinImageTask.java
package com.chess.cryptobot.task;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.AsyncTask;
import android.util.Log;

import androidx.annotation.Nullable;

import com.chess.cryptobot.content.balance.BalanceHolder;
import com.chess.cryptobot.model.Balance;

import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.lang.ref.WeakReference;
import java.net.URL;

public class CoinImageTask extends AsyncTask<Balance, Integer, Balance> {
    private final WeakReference<BalanceHolder> balanceHolderWeakReference;
    private final static String TAG = "CoinImageTask";

    public CoinImageTask(BalanceHolder balanceHolder) {
        this.balanceHolderWeakReference = new WeakReference<>(balanceHolder);
    }

    @Override
    protected Balance doInBackground(Balance... balances) {
        boolean updated = false;
        Balance balance = balances[0];
        try {
            Bitmap bitmap = getImage(balance);
            balance.setCoinIcon(bitmap);
            updated = true;
        } catch (IOException ignored) {
        }

        if (!updated) {
            cancel(true);
            return null;
        }
        return balance;
    }

    @Override
    protected void onPostExecute(@Nullable Balance balance) {
        if (balance == null) return;
        BalanceHolder balanceHolder = balanceHolderWeakReference.get();
        if (balanceHolder != null) balanceHolder.setItem(balance);
    }

    private Bitmap getImage(Balance balance) throws IOException {
        Bitmap bitmap;
        String coinName = balance.getName();
        try {
            bitmap = loadImage(coinName);
        } catch (IOException e) {
            bitmap = downloadImage(balance.getCoinUrl());
            saveImage(bitmap, fileName(coinName));
        }

        return bitmap;
    }

    private Bitmap loadImage(String coinName) throws IOException {
        Context context = getContext();
        if (context == null) return null;
        try (FileInputStream fileInputStream = context.openFileInput(fileName(coinName))) {
            return BitmapFactory.decodeStream(fileInputStream);
        }
    }

    private String fileName(String coinName) {
        return String.format("%s.png", coinName.toLowerCase());
    }

    private void saveImage(Bitmap bitmap, String fileName) {
        Context context = getContext();
        if (context == null || bitmap == null) return;
        try (FileOutputStream out = context.openFileOutput(fileName, Context.MODE_PRIVATE)) {
            bitmap.compress(Bitmap.CompressFormat.PNG, 100, out);
        } catch (IOException e) {
            Log.d(TAG, e.getLocalizedMessage());
        }
    }

    private Bitmap downloadImage(String coinUrl) throws IOException {
        URL imageUrl = new URL(coinUrl);
        try (InputStream inputStream = imageUrl.openStream()) {
            return BitmapFactory.decodeStream(inputStream);
        }
    }

    private Context getContext() {
        BalanceHolder balanceHolder = balanceHolderWeakReference.get();
        if (balanceHolder != null) {
            return balanceHolder.getContext();
        } else {
            cancel(true);
            return null;
        }
    }
}

File: app/src/main/java/com/chess/cryptobot/task/CoinImageTask.kt
package com.chess.cryptobot.task

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Log
import com.chess.cryptobot.content.balance.BalanceHolder
import com.chess.cryptobot.model.Balance
import kotlinx.coroutines.*
import kotlinx.coroutines.Dispatchers.IO
import java.io.IOException
import java.net.URL
import java.util.*
import java.util.concurrent.Executors

class CoinImageTask(private val balanceHolder: BalanceHolder) {
    private val executor = Executors.newSingleThreadExecutor()
    private val scope = CoroutineScope(SupervisorJob() + executor.asCoroutineDispatcher())
    private val tag = CoinImageTask::class.qualifiedName

    fun doInBackground(vararg params: Balance) {
        val balance = params[0]

        try {
            scope.launch(IO) {
                val bitmap = getImage(balance)
                balance.coinIcon = bitmap
                withContext(Dispatchers.Main) {
                    onPostExecute(balance)
                }
            }
        } catch (e: IOException) {
            Log.e(tag, Log.getStackTraceString(e))
        }

    }

    private fun onPostExecute(balance: Balance?) {
        if (balance == null) return
        balanceHolder.setItem(balance)
    }

    @Throws(IOException::class)
    private fun getImage(balance: Balance): Bitmap? {
        var bitmap: Bitmap?
        val coinName = balance.name
        try {
            bitmap = loadImage(coinName)
        } catch (e: IOException) {
            if (balance.coinUrl == null) return null
            bitmap = downloadImage(balance.coinUrl)
            saveImage(bitmap, fileName(coinName))
        }
        return bitmap
    }

    @Throws(IOException::class)
    private fun loadImage(coinName: String): Bitmap? {
        context.openFileInput(fileName(coinName))
            .use { fileInputStream -> return BitmapFactory.decodeStream(fileInputStream) }
    }

    private fun fileName(coinName: String): String {
        return String.format("%s.png", coinName.lowercase(Locale.getDefault()))
    }

    private fun saveImage(bitmap: Bitmap?, fileName: String) {
        if (bitmap == null) return
        try {
            context.openFileOutput(fileName, Context.MODE_PRIVATE)
                .use { out -> bitmap.compress(Bitmap.CompressFormat.PNG, 100, out) }
        } catch (e: IOException) {
            Log.e(tag, Log.getStackTraceString(e))
        }
    }

    @Throws(IOException::class)
    private fun downloadImage(coinUrl: String?): Bitmap {
        val imageUrl = URL(coinUrl)
        imageUrl.openStream().use { inputStream -> return BitmapFactory.decodeStream(inputStream) }
    }

    private val context: Context
        get() {
            return balanceHolder.context
        }
}